### PR TITLE
Adding retries to AWS API calls

### PR DESCRIPTION
**Pull request by biacar**
## Proposed Changes

This PR addresses the issues related to transient errors while querying AWS APIs. Retries have been added to AWS API calls with a sleep time in between the calls to make sure whatever the error code returned by AWS is not transient. Credential errors are not retried.

The motivation behind this PR is unlock failures during peer discovery due to timed out AWS calls. The retries on the AWS calls would reduce the possibility of a node not releasing the discovery lock after successfully joining a cluster.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [ x ] I have read the `CONTRIBUTING.md` document
- [ x ] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ x ] All tests pass locally with my changes
- [ x ] I have added tests that prove my fix is effective or that my feature works
- [ x ] I have added necessary documentation (if appropriate)
- [ x ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

File: deps/rabbitmq_aws/src/rabbitmq_aws.erl [hunk 1]
```diff
@@ -543,10 +543,22 @@ ensure_credentials_valid() ->
...
+  %% @end
+api_get_request_with_retries(Service, Path, Retries, WaitTimeBetweenRetries) ->
   ensure_credentials_valid(),
   case get(Service, Path) of
     {ok, {_Headers, Payload}} -> rabbit_log:debug("AWS request: ~s~nResponse: ~p", [Path, Payload]),
                                  {ok, Payload};
     {error, {credentials, _}} -> {error, credentials};
-    {error, Message, _}       -> {error, Message}
+    {error, Message, _}       -> case Retries of
+                                   0 -> {error, Message};
```
Comment (michaelklishin) [inline deps/rabbitmq_aws/src/rabbitmq_aws.erl:559]: It's more common to match on `0` retries left in a function head, here is [an example](https://github.com/rabbitmq/rabbitmq-server/blob/master/deps/rabbit/src/rabbit_mnesia.erl#L94-L103).

**Review by michaelklishin — CHANGES_REQUESTED** [review]

File: deps/rabbitmq_aws/src/rabbitmq_aws.erl (hunk 1 above)
Comment (biacar) [reply deps/rabbitmq_aws/src/rabbitmq_aws.erl]: Thank you for the suggestion, done!

**Comment by michaelklishin** [comment]
Merged manually.

**Comment by michaelklishin** [comment]
Backported to `v3.9.x`.

**Comment by michaelklishin** [comment]
Two cosmetic follow-up PRs: https://github.com/rabbitmq/rabbitmq-server/commit/59ebb83506f21ba70ada1894b6817324c71eef24, https://github.com/rabbitmq/rabbitmq-server/commit/6bf4a99f58dcd22107da1a232ab98e86249b2066

**Pull request closed without merging**